### DISCUSSION TITLE
Run mkosi.images subimage using nspawn or qemu

### DISCUSSION THREAD

**Discussion by slarew** [systemd/mkosi#3150] [Q&A] [answered]
I have a main image defined in `project/mkosi.conf` and several subimages it depends on in `project/mkosi.images/{image_A.conf,image_B.conf,image_C.conf}`. Is it possible to use the `mkosi boot` (aka nspawn) or `mkosi qemu` verbs to run a sub-image produced from `mkosi.images`?

I'm thinking something like:
```
# Run image_B in qemu:
mkosi --run-image-id image_B qemu
```

**Comment by slarew** [comment]
I found a possible solution to achieve my goal of running a subimage. I "claim" the subimage as the main image by copying relevant fields from `[Output]` in the subimage into the `[Output]` in the main image.

e.g. `project/mkosi.images/image_B.conf`:
```
[Output]
# Just for this example, be explicit on the format and output so they can be matched in the main image:
Format=disk
Output=image_b
```
and `project/mkosi.conf`:
```
[Output]
# Match the format and output name to the subimage image_B:
Format=disk
Output=image_B
```
Then I can run image_B in qemu:
```
cd project
mkosi qemu
```

**Reply by slarew** [reply]
I figured out that I can do this by reading the mkosi code, specifically `qemu.py` and backing into the solution. It feels like I'm taking advantage of undefined behavior (to borrow the term) for this solution. Is this a fragile approach that is prone to breaking?

**Reply by daandemeyer** [reply, marked as answer]
> I figured out that I can do this by reading the mkosi code, specifically `qemu.py` and backing into the solution. It feels like I'm taking advantage of undefined behavior (to borrow the term) for this solution. Is this a fragile approach that is prone to breaking?

Yes we can't guarantee this will keep working. Unfortunately what you're asking for is not supported, mkosi will only boot the top level image, mkosi.images is intended for subimages intended for inclusion or use with the top level image that are not standalone and cannot be booted.

If you want to build multiple images that can all be booted, we recommend having two separate projects and using `mkosi -C` to boot a specific one. `Include=` can be used if configuration needs to be shared across the two projects.

**Reply by slarew** [reply]
OK thanks for clarifying that's helpful. I will change my approach.
